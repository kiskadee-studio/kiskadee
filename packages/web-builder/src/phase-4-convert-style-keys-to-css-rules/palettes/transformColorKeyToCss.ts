import {
  type ColorProperty,
  CssColorProperty,
  classNameCssPseudoSelector,
  type HSLA,
  InteractionStateCssPseudoSelector,
  type StyleKey
} from '@kiskadee/schema';
import { convertHslaToHex } from '../utils/convertHslaToHex';

export const ERROR_INVALID_KEY_FORMAT =
  'Invalid key format. Expected value in square brackets at the end.';
export const ERROR_REF_REQUIRE_STATE =
  'Invalid key format. Reference "==" requires a preceding non-rest interaction state.';

/**
 * Transform a style key into its corresponding CSS rule representation.
 *
 * Handles two cases:
 * 1. Simple keys (no "=="): generates a class rule, optionally with a pseudo-state
 *    if the key contains an interaction state (e.g. "--hover__").
 * 2. Reference keys ("=="): generate a parent rule with a pseudo-state and
 *    a nested rule targeting the child selector. Also supports legacy "::ref".
 *
 * @param styleKey - the style token, e.g. "boxColor--hover__[240,50,50,0.5]" or
 *                   "boxColor==hover__[240,50,50,0.5]"
 * @param className - the CSS class name to use for the generated rule (without the leading dot).
 *                    Usually this is the shortened token assigned to the styleKey (for example "a"
 *                    or "abc"). The function will emit selectors using `.${className}`.
 * @param forceState - when true and a pseudo-class is present, also include the corresponding
 *                     "forced" CSS class selector (from classNameCssPseudoSelector) alongside
 *                     the native pseudo-class so the same style can be applied by adding that
 *                     class in HTML. Example: if pseudo-class is "hover" and its forced suffix
 *                     is "-h", the generated selector list will include ".abc:hover, .abc.-h" for
 *                     inline rules or ".-a:hover .abc, .-a.-h .abc" for parent-ref rules.
 * @returns GeneratedCss containing:
 *   - className: token without a dot prefix, for use in HTML
 *   - cssRule: full CSS text including selector(s)
 *   - parentClassName: only for reference cases (== or ::ref), the full token for the parent selector
 */
export function transformColorKeyToCss(
  styleKey: StyleKey,
  className: string,
  forceState?: boolean
): string {
  // Extract HSLA values from brackets at the end; error if missing.
  const hslaMatch = styleKey.match(/\[([^\]]+)]$/);
  const hasHslaMatch = hslaMatch !== null;
  if (hasHslaMatch === false) {
    throw new Error(ERROR_INVALID_KEY_FORMAT);
  }
  const hsla = hslaMatch[1].split(',').map(Number) as HSLA;
  const hex = convertHslaToHex(hsla);

  // Base color property, e.g. "background-color" or "color"
  // Support both non-ref ("--" or "__") and ref ("==") separators
  const propertyName = styleKey.split(/==|--|__/)[0] as ColorProperty;
  const colorProperty = CssColorProperty[propertyName];

  // Prepare interaction state patterns using known state keys rather than pseudo map,
  // so we can support states that intentionally have no native pseudo (e.g., disabled).
  const stateKeys = ['hover', 'pressed', 'selected', 'focus', 'disabled', 'readOnly'];
  const inlinePseudoClassRegex = new RegExp(`--(${stateKeys.join('|')})(?=__)`);
  const newRefStateOnChildRegex = new RegExp(`==(${stateKeys.join('|')})(?=$)`);

  let cssRule: string;

  const isRef = styleKey.includes('==');
  const hasRef = isRef === true;

  if (hasRef === false) {
    // Simple class may include a pseudo-state
    const match = styleKey.match(inlinePseudoClassRegex);
    const pseudoClass = match ? (match[1] as string) : undefined;
    const hasPseudoClass = pseudoClass !== undefined;
    if (hasPseudoClass === true) {
      // build selectors list: prefer native pseudo when available; use forced class for disabled
      const selectors: string[] = [];

      const nativePseudo = (InteractionStateCssPseudoSelector as Record<string, string>)[
        pseudoClass
      ] as string | undefined;
      if (nativePseudo && nativePseudo !== '') {
        selectors.push(`.${className}${nativePseudo}`);
      }

      const forcedSuffixFor = (state: string): string => {
        return (classNameCssPseudoSelector as Record<string, string>)[state] ?? '';
      };
      const forcedSuffix = forcedSuffixFor(pseudoClass);
      const hasForcedSuffix = forcedSuffix !== '';

      const shouldAddForced =
        pseudoClass === 'disabled' || (forceState === true && hasForcedSuffix);
      if (shouldAddForced && hasForcedSuffix) {
        // forcedSuffix already contains the short token (e.g. "-h"); apply it gated by activator "-a" on the same element
        // Requirement: the forced state class alone must NOT activate styles; it must be combined with "-a".
        selectors.push(`.${className}.${forcedSuffix}.-a`);
      }

      const selector = selectors.join(', ');
      cssRule = `${selector} { ${colorProperty}: ${hex} }`;
    } else {
      cssRule = `.${className} { ${colorProperty}: ${hex} }`;
    }
    return cssRule;
  }

  // Reference pseudo-class: split out child selector
  const [child] = styleKey.split('__');

  let pseudoClass: string | undefined;
  const match = child.match(newRefStateOnChildRegex);
  pseudoClass = match ? match[1] : undefined;

  const isInvalidState = pseudoClass === undefined || pseudoClass === '';

  if (isInvalidState === true) {
    throw new Error(ERROR_REF_REQUIRE_STATE);
  }

  // parent selector with native pseudo-class and optional forced-class variant
  const parentSelectors: string[] = [];
  const nativePseudo = (InteractionStateCssPseudoSelector as Record<string, string>)[
    // biome-ignore lint/style/noNonNullAssertion: ...
    pseudoClass!
  ] as string | undefined;
  if (nativePseudo && nativePseudo !== '') {
    parentSelectors.push(`.-a${nativePseudo} .${className}`);
  }

  const forcedSuffixFor = (state: string): string => {
    return (classNameCssPseudoSelector as Record<string, string>)[state] ?? '';
  };
  // biome-ignore lint/style/noNonNullAssertion: ...
  const forcedSuffix = forcedSuffixFor(pseudoClass!);
  const hasForcedSuffix = forcedSuffix !== '';

  const shouldAddForcedRef = pseudoClass === 'disabled' || (forceState === true && hasForcedSuffix);
  if (shouldAddForcedRef && hasForcedSuffix) {
    // forced class applied on the parent (parent will have both -a and forcedSuffix classes), e.g. ".-a.-h .abc"
    parentSelectors.push(`.-a.${forcedSuffix} .${className}`);
  }

  const selector = parentSelectors.join(', ');
  return `${selector} { ${colorProperty}: ${hex} }`;
}
