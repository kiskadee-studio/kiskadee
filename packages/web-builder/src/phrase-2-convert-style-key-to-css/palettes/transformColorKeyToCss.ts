import {
  type ColorKeys,
  CssColorProperty,
  type HLSA,
  InteractionStateCssMapping
} from '@kiskadee/schema';
import { convertHslaToHex } from '../utils/convertHslaToHex';
import type { GeneratedCss } from '../phrase2.types';

export const ERROR_INVALID_KEY_FORMAT =
  'Invalid key format. Expected value in square brackets at the end.';
export const ERROR_REF_REQUIRE_STATE =
  'Invalid key format. "::ref" requires a preceding interaction state.';

/**
 * Transform a style key into its corresponding CSS rule representation.
 *
 * Handles two cases:
 * 1. Simple keys (no "::ref"): generates a class rule, optionally with a pseudo-state
 *    if the key contains an interaction state (e.g. "--hover__").
 * 2. Reference keys ("::ref"): generates a parent rule with a pseudo-state and
 *    a nested rule targeting the child selector.
 *
 * @param styleKey - the style token, e.g. "bgColor--hover__[240,50,50,0.5]" or
 *                   "bgColor--hover::ref__[240,50,50,0.5]"
 * @returns GeneratedCss containing:
 *   - className: token without dot prefix, for use in HTML
 *   - cssRule: full CSS text including selector(s)
 *   - parentClassName: only for ::ref cases, the full token for the parent selector
 */
export function transformColorKeyToCss(styleKey: string): GeneratedCss {
  // Extract HSLA values from brackets at end; error if missing.
  const hslaMatch = styleKey.match(/\[([^\]]+)]$/);
  if (hslaMatch === null) {
    throw new Error(ERROR_INVALID_KEY_FORMAT);
  }
  const hsla = hslaMatch[1].split(',').map(Number) as HLSA;
  const hex = convertHslaToHex(hsla);

  // Base color property, e.g. "background-color" or "color"
  const colorKey = styleKey.split(/--|__/)[0] as ColorKeys;
  const cssProp = CssColorProperty[colorKey];

  // Prepare interaction state patterns (e.g. "hover", "focus", ...)
  const states = Object.values(InteractionStateCssMapping).map((s) => s.slice(1));
  const inlineStateRegex = new RegExp(`--(${states.join('|')})(?=__)`);
  const refStateRegex = new RegExp(`--(${states.join('|')})(?=::ref)`);

  let className: string;
  let parentClassName: string | undefined;
  let cssRule: string;

  // Determine if this is a reference style
  if (!styleKey.includes('::ref')) {
    // Simple class, may include a pseudo-state
    className = styleKey;
    const match = styleKey.match(inlineStateRegex);
    if (match !== null) {
      cssRule = `.${styleKey}:${match[1]} { ${cssProp}: ${hex}; }`;
    } else {
      cssRule = `.${styleKey} { ${cssProp}: ${hex}; }`;
    }
  } else {
    // Reference state: split out child selector
    const [child] = styleKey.split('__');
    className = child;
    const match = child.match(refStateRegex);
    if (match === null) {
      throw new Error(ERROR_REF_REQUIRE_STATE);
    }
    parentClassName = styleKey;
    cssRule = `.${styleKey}:${match[1]} .${child} { ${cssProp}: ${hex}; }`;
  }

  return { className, parentClassName, cssRule };
}
