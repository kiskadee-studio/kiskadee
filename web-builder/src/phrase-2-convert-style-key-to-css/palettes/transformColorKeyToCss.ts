import { type ColorKeys, CssColorProperty } from '@kiskadee/schema';
import { convertHslaToHex } from '../utils/convertHslaToHex';

/**
 * Transforms a key generated by convertPalettesToKeys into a CSS rule.
 *
 * Rules:
 * - If the key does NOT contain "::ref", output is:
 *
 *     .[key] {
 *       [cssProperty]: <hex-converted-color>;
 *     }
 *
 * - If the key contains "::ref", output is two selectors:
 *
 *     .[fullKey] .[refSelector] {
 *       [cssProperty]: <hex-converted-color>;
 *     }
 *
 * The CSS property is determined from the palette key using the cssPropertyMap.
 */
export function transformColorKeyToCss(key: string): string {
  // Regular expression to capture the value in square brackets at the end of the string
  const valueRegex = /\[([^\]]+)]$/;
  const valueMatch = key.match(valueRegex);

  if (!valueMatch) {
    throw new Error('Invalid key format. Expected value in square brackets at the end.');
  }

  // Extract the raw color value (e.g., "[120,50,50,1]") and then extract the numeric values
  const valueStr = valueMatch[1];
  const hsla = valueStr.split(',').map((v) => Number.parseFloat(v)) as [
    number,
    number,
    number,
    number
  ];

  // Convert HSLA value to hexadecimal color
  const hexColor = convertHslaToHex(hsla);

  // Determine the palette key (e.g., "textColor") that maps to a CSS property.
  // Splits on "--" or "__" to get the palette property.
  const colorKey = key.split(/--|__/)[0];
  const cssProperty = CssColorProperty[colorKey as ColorKeys];

  const hasRef = key.includes('::ref');
  let cssRule = '';

  if (hasRef) {
    // When the key includes "::ref", the referential selector is the key without the "__[value]" part.
    const [refSelector] = key.split('__');
    cssRule = `.${key} .${refSelector} { ${cssProperty}: ${hexColor}; }`;
  } else {
    cssRule = `.${key} { ${cssProperty}: ${hexColor}; }`;
  }

  return cssRule;
}
