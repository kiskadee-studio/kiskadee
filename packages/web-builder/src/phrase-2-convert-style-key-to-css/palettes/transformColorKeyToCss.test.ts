import { describe, it, expect } from 'vitest';
import { transformColorKeyToCss } from './transformColorKeyToCss';
import type { GeneratedCss } from '../phrase2.types';

describe('transformColorKeyToCss', () => {
  describe('Success operation', () => {
    it('should transform a key without "::ref"', () => {
      const styleKey = 'textColor__[120,50,50,1]';
      const rule = '.textColor__[120,50,50,1] { color: #40bf40; }';
      const expected: GeneratedCss = {
        className: styleKey,
        cssRule: rule
      };
      expect(transformColorKeyToCss(styleKey)).toEqual(expected);
    });

    it('should transform a key with "--hover::ref" and include :hover on parent', () => {
      const result = transformColorKeyToCss('bgColor--hover::ref__[240,50,50,0.5]');

      const expected: GeneratedCss = {
        className: 'bgColor--hover::ref',
        parentClassName: 'bgColor--hover::ref__[240,50,50,0.5]',
        cssRule:
          '.bgColor--hover::ref__[240,50,50,0.5]:hover .bgColor--hover::ref { background-color: #4040bf80; }'
      };

      expect(result).toEqual(expected);
      expect(result).toMatchSnapshot();
    });

    it('should transform a key without "::ref" and include ":hover"', () => {
      const result = transformColorKeyToCss('bgColor--hover__[240,50,50,0.5]');

      const expected: GeneratedCss = {
        className: 'bgColor--hover__[240,50,50,0.5]',
        cssRule: '.bgColor--hover__[240,50,50,0.5]:hover { background-color: #4040bf80; }',
        parentClassName: undefined
      };

      expect(result).toEqual(expected);
      expect(result).toMatchSnapshot();
    });

    it('should transform a key with "--focus::ref" and include :focus on parent', () => {
      const key = 'textColor--focus::ref__[0,0,0,0.3]';
      const child = 'textColor--focus::ref';
      const rule =
        '.textColor--focus::ref__[0,0,0,0.3]:focus .textColor--focus::ref { color: #0000004d; }';
      const expected: GeneratedCss = {
        className: child,
        parentClassName: key,
        cssRule: rule
      };
      expect(transformColorKeyToCss(key)).toEqual(expected);
    });
  });

  describe('Error handling', () => {
    it('should throw if "::ref" is used without a state', () => {
      const key = 'bgColor::ref__[240,50,50,0.5]';
      expect(() => transformColorKeyToCss(key)).toThrowError(
        'Invalid key format. "::ref" requires a preceding interaction state.'
      );
    });

    it('should throw if the key format is invalid', () => {
      expect(() => transformColorKeyToCss('invalidKey')).toThrowError(
        'Invalid key format. Expected value in square brackets at the end.'
      );
    });

    it('should throw when using unsupported state "visited"', () => {
      const key = 'bgColor--visited::ref__[240,50,50,0.5]';
      expect(() => transformColorKeyToCss(key)).toThrowError(
        'Invalid key format. "::ref" requires a preceding interaction state.'
      );
    });
  });
});
