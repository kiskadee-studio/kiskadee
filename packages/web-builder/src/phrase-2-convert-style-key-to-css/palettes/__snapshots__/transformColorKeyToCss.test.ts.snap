// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformColorKeyToCss > Error handling > should throw if "::ref" is used without a state 1`] = `[Error: Invalid key format. "::ref" requires a preceding interaction state.]`;

exports[`transformColorKeyToCss > Error handling > should throw if the style key format is invalid 1`] = `[Error: Invalid key format. Expected value in square brackets at the end.]`;

exports[`transformColorKeyToCss > Error handling > should throw when using unsupported state "visited" 1`] = `[Error: Invalid key format. "::ref" requires a preceding interaction state.]`;

exports[`transformColorKeyToCss > Success operation > should transform a key with "--hover::ref" and include :hover on parent 1`] = `
{
  "className": "bgColor--hover::ref",
  "cssRule": ".bgColor--hover::ref__[240,50,50,0.5]:hover .bgColor--hover::ref { background-color: #4040bf80; }",
  "parentClassName": "bgColor--hover::ref__[240,50,50,0.5]",
}
`;

exports[`transformColorKeyToCss > Success operation > should transform a key without "::ref" and include ":hover 1`] = `
{
  "className": "bgColor--hover__[240,50,50,0.5]",
  "cssRule": ".bgColor--hover__[240,50,50,0.5] { background-color: #4040bf80; }",
  "parentClassName": undefined,
}
`;

exports[`transformColorKeyToCss > Success operation > should transform a key without "::ref" and include ":hover" 1`] = `
{
  "className": "bgColor--hover__[240,50,50,0.5]",
  "cssRule": ".bgColor--hover__[240,50,50,0.5]:hover { background-color: #4040bf80; }",
}
`;
